modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -3770,41 +3792,63 @@ def _groupby_shuffle(
-            by = [by]
+        if any(
+            isinstance(obj, pandas.Grouper)
+            for obj in (by if isinstance(by, list) else [by])
+        ):
+            raise NotImplementedError(
+                "Grouping on a pandas.Grouper with range-partitioning groupby is not yet supported: "
+                + "https://github.com/modin-project/modin/issues/5926"
+            )
 
-        is_all_labels = all(isinstance(col, (str, tuple)) for col in by)
-        is_all_column_names = (
-            all(col in self.columns for col in by) if is_all_labels else False
+        external_by, internal_by, by_positions = self._groupby_internal_columns(
+            by, drop
         )
 
-        is_transform = how == "transform" or GroupBy.is_transformation_kernel(agg_func)
+        all_external_are_qcs = all(isinstance(obj, type(self)) for obj in external_by)
+        if not all_external_are_qcs:
+            raise NotImplementedError(
+                "Grouping on an external grouper with range-partitioning groupby is only supported with Series'es: "
+                + "https://github.com/modin-project/modin/issues/5926"
+            )
 
+        is_transform = how == "transform" or GroupBy.is_transformation_kernel(agg_func)
         if is_transform:
             # https://github.com/modin-project/modin/issues/5924
             ErrorMessage.missmatch_with_pandas(
                 operation="range-partitioning groupby",
                 message="the order of rows may be shuffled for the result",
             )
 
-        if not is_all_column_names:
-            raise NotImplementedError(
-                "Range-partitioning groupby is only supported when grouping on a column(s) of the same frame. "
-                + "https://github.com/modin-project/modin/issues/5926"
-            )
-
         # This check materializes dtypes for 'by' columns
         if not is_transform and groupby_kwargs.get("observed", False) in (
             False,
             lib.no_default,
         ):
-            if isinstance(self._modin_frame._dtypes, ModinDtypes):
-                by_dtypes = self._modin_frame._dtypes.lazy_get(by).get()
-            else:
-                by_dtypes = self.dtypes[by]
+            # The following 'dtypes' check materializes dtypes for 'by' columns
+            internal_dtypes = pandas.Series()
+            external_dtypes = pandas.Series()
+            if len(internal_by) > 0:
+                internal_dtypes = (
+                    self._modin_frame._dtypes.lazy_get(internal_by).get()
+                    if isinstance(self._modin_frame._dtypes, ModinDtypes)
+                    else self.dtypes[internal_by]
+                )
+            if len(external_by) > 0:
+                for obj in external_by:
+                    if not isinstance(obj, type(self)):
+                        # we're only interested in categorical dtypes here, which can only
+                        # appear in pandas objects
```

Comment:
?
```suggestion
                        # appear in Modin objects
```

Reply:
Could you also add test for the case?

Reply:
we won't be able to hit this line with tests, as all non-modin objects [are filtered out here](https://github.com/dchigarev/modin/blob/4d3e41819559c311518a866fdde43a10cad841a6/modin/core/storage_formats/pandas/query_compiler.py#L3814-L3819), I added this safeguard so we wouldn't hit the error in the future (when the support for non-modin groupers will be enabled) trying to access `.dtypes` field of non-modin objects